R/drive_read.R

```diff
@@ -0,0 +1,298 @@
+#'      drive_upload(name = "r_logo.jpg")
+#'   img <- r_logo %>%
+#'     drive_read() %>%
+#'     jpeg::readJPEG()
+#'   plot(0:1, 0:1, type = "n")
+#'   rasterImage(img, 0, 0, 1, 1)
+#'
+#'   # JPEG cleanup
+#'   drive_rm(r_logo)
+#' }
+#'
+#' # html -> character vector
+#' # https://stat.ethz.ch/R-manual/R-patched/doc/html/about.html
+#' if (require(rvest) && require(readr)) {
+#'   r_about <- R.home("doc/html/about.html") %>%
+#'      drive_upload()
+#'   r_about %>%
+#'     drive_read() %>%
+#'     rvest::html_text2() %>%
+#'     readr::read_lines() %>%
+#'     lapply(substr, start = 1, stop = 70)
+#'
+#'   # html clean up
+#'   drive_rm(r_about)
+#' }
+#'
+#' # html --> character vector
+#' if (require(rvest) && require(stringi)) {
+#'   tfile <- tempfile(fileext = ".html")
+#'   curl::curl_download("https://httpbin.org/html", destfile = tfile)
+#'   httpbin_html <- tfile %>%
+#'     drive_upload(name = "httpbin.html")
+#'   httpbin_html %>%
+#'     drive_read() %>%
+#'     rvest::html_text2() %>%
+#'     stringi::stri_split_boundaries(type = "sentence") %>%
+#'     lapply(substr, start = 1, stop = 70) %>%
+#'     .[[1]] %>%
+#'     head(6)
+#'
+#'   # html clean up
+#'   drive_rm(httpbin_html)
+#' }
+#'
+#' # xml --> list
+#' if (require(xml2)) {
+#'   tfile <- tempfile(fileext = ".xml")
+#'   curl::curl_download("https://httpbin.org/xml", destfile = tfile)
+#'   httpbin_xml <- tfile %>%
+#'     drive_upload()
+#'   httpbin_xml %>%
+#'     drive_read() %>%
+#'     xml2::as_list()
+#'
+#'   # xml clean up
+#'   drive_rm(httpbin_xml)
+#' }
+#'
+#' # json --> list
+#' tfile <- tempfile(fileext = ".json")
+#' curl::curl_download("https://httpbin.org/json", destfile = tfile)
+#' httpbin_json <- tfile %>%
+#'   drive_upload()
+#' httpbin_json %>%
+#'   drive_read()
+#'
+#' # json clean up
+#' drive_rm(httpbin_json)
+drive_read <- function(file,
+                       type = NULL,
+                       as = c("auto", "string", "raw", "json", "html", "xml"),
+                       ...) {
+  as <- match.arg(as)
+  file <- as_dribble(file)
+  file <- confirm_single_file(file)
+
+  mime_type <- pluck(file, "drive_resource", 1, "mimeType")
+
+  if (!grepl("google", mime_type) && !is.null(type)) {
+    drive_bullets(c(
+      "!" = "Ignoring {.arg type}. Only consulted for native Google file types.",
+      " " = "MIME type of {.arg file}: {.field {mime_type}}."
+    ))
+  }
+
+  if (grepl("google", mime_type)) {
+    export_type <- type %||% get_export_mime_type(mime_type)
+    export_type <- drive_mime_type(export_type)
+    verify_export_mime_type(mime_type, export_type)
+
+    request <- request_generate(
+      endpoint = "drive.files.export",
+      params = list(
+        fileId = file$id,
+        mimeType = export_type
+      )
+    )
+  } else {
+    request <- request_generate(
+      endpoint = "drive.files.get",
+      params = list(
+        fileId = file$id,
+        alt = "media"
+      )
+    )
+  }
```

Comment:
This is duplicated with code in `drive_download()`. But since it's 2x and not yet 3x, I think that's OK. It would be a weird bit of code to factor out into a function.